### PR TITLE
Fix wallet scheme value checks

### PR DESCRIPTION
**Pull request by AugustoL**
- Fix check of value transferred in proposals, now the total value sent by asset and receiver is checked on the permission. Example if permission to transfer DAI to address A is 100 DAI and a proposal with two DAI transfer is created (one with 50 and the other with 51) the proposal will revert execution.
- Fix check of value transferred in ERC20 transfers. Previously we were not checking if there was ETH transferred in ERC20 transfers, this could have caused sending ETH to ERC20 token contracts.

Updated WalletScheme tokens to reach 100% test coverage, only was require statement was left unchecked, but we I think it is imposible for it to happen, anyways we can consider full WalletScheme contract coverage in statements and branch levels.

### CODE REVIEW

**Review by blackscale-ethsec — COMMENTED** [review]
https://github.com/AugustoL/dxdao-contracts/pull/2/commits/312cbcab29f9db4de15d61aea56649a979c4e474

L.120 comment does not match code
```
// If the amount of time passed since submission plus max proposal time is higher than block timestamp
        // the proposal timeout execution is reached and proposal cant be executed from now on
        if (proposal.submittedTime.add(maxProposalTime) < now) {
```
pseudo code inferred from comment: ```if (secondsSinceSubmission + maxProposalTime > blockTimestamp) {make proposal unexecutable}```

... which would not work since we would get the sum of 2 relatively short "seconds-elapsed", that can never be higher than a timestamp (which is seconds elapsed since 1.1.1970)

Switching out secondsSinceSubmission for submittedTime (a timestamp), gets us closer to a solution, but that means the comment should read "If the timestamp of the proposal submission + maxProposalTime ...".
The inequality is also wrongly stated in the comment.
The code correctly states that if the current timestamp is higher than SUM, then SUM is the past and the deadline is over, and since SUM = executionDeadline, we are correctly executing the block in the right situation.

Suggestions: 
- correct comment to match code
- for clarity, consider adapting nomenclature to use -Time suffix only with timestamps, and use -Seconds suffix or prefix to declare "time-elapsed" or duration variables.

**Review by blackscale-ethsec — COMMENTED** [review]
https://github.com/AugustoL/dxdao-contracts/pull/2/commits/312cbcab29f9db4de15d61aea56649a979c4e474

L.123

I havent fully grasped the entirety of the code yet, so it might just be me missing a key piece of the puzzle.
I noticed that the change of state for the proposal happens after the require of L. 118, which is supposed to protect a proposal from undue execution.
If there is no other protection mechanism, an over-the-deadline proposal could go through anyway (even while emitting a ProposalExecutionTimeoutEvent)

**Comment by blackscale-ethsec** [comment]
`maxSecondsForExecution` looks unambiguous and clear

**Review by blackscale-ethsec — COMMENTED** [review]
you could also tag the test with @skip-on-coverage and instruct solcover to skip those
cf. https://github.com/sc-forks/solidity-coverage/blob/master/docs/advanced.md

**Review by blackscale-ethsec — COMMENTED** [review]
https://github.com/AugustoL/dxdao-contracts/pull/2/commits/4eb4bf52baaadfab8f12f4467bd315ee3a804d65

L.256

scratch that, I need new glasses, misread the inequality.

```
if (callDataFuncSignature == 
  ERC20_TRANSFER_SIGNATURE)
  require(_value[i] == 0, "cant propose ERC20 trasnfers with value");

vs.

require(callDataFuncSignature != ERC20_TRANSFER_SIGNATURE || _value[i] == 0, "cant propose ERC20 trasnfers with value");
```
(small typo on transfers btw)

removed lines say:
if calldata contains a transfer method sig, then value (eth sent along) must be zero, or revert

The proposed replacement code says:
calldata must contain a transfer sig, OR, failing that, eth sent along must be zero.
So basically, I could send eth along with an erc20 transfer, but i can t send any eth if the calldata refers to any other method (or none).

Refactor is changing functionality. Tests didnt catch it, maybe a new test is needed ?
I havent had time to go through those yet, maybe there was one but it needs tweaks to account for all situations ?

**Comment by AugustoL** [comment]
condition 1(e.g., X) | condition 2(e.g., Y) | NOT X( ~ X ) | X AND Y( X && Y ) | X OR Y( X \|\| Y )
-- | -- | -- | -- | --
false | false | true | false | false
false | true | true | false | true
true | false | false | false | true
true | true | false | true | true

<br class="Apple-interchange-newline">

With the double false OR condition check we make sure that it throws only in the case where it is a transfer and value is being sent, this is the only way I found to have the require in one line condition :grimacing:

**Comment by blackscale-ethsec** [comment]
what about https://github.com/AugustoL/dxdao-contracts/pull/2#pullrequestreview-655953988
?

**Comment by AugustoL** [comment]
> what about [#2 (review)](https://github.com/AugustoL/dxdao-contracts/pull/2#pullrequestreview-655953988)
> ?

That is the signature of the setMaxSecondsForExecution function, the only function that we allowed to be proposed using the walletscheme as receiver (the wallet scheme calls "itself" to set that variable). The first time I added it was hadcoded bytes4 of the signature in the condition, I thought it would be better to have it as a constant with a clear name in the code.
